File: src/managers/scheduler.js
import fs from 'fs';

const SCHEDULES_FILE = 'schedules.json';

/**
 * Scheduler Manager
 */
export class Scheduler {
    constructor() {
        this.schedules = [];
        this.loadSchedules();
    }

    loadSchedules() {
        if (fs.existsSync(SCHEDULES_FILE)) {
            try {
                const data = fs.readFileSync(SCHEDULES_FILE, 'utf8');
                const parsed = JSON.parse(data);
                this.schedules = parsed.schedules || [];
            } catch (error) {
                console.error('⚠️  Error loading schedules:', error);
                this.schedules = [];
            }
        }
    }

    saveSchedules() {
        try {
            fs.writeFileSync(SCHEDULES_FILE, JSON.stringify({ schedules: this.schedules }, null, 2), 'utf8');
        } catch (error) {
            console.error('⚠️  Error saving schedules:', error);
        }
    }

    createSchedule(userId, channelId, scheduledTime, recurring = null) {
        const scheduleId = this.schedules.length + 1;
        
        // Parse scheduled time
        let scheduledDate;
        if (scheduledTime.startsWith('+')) {
            // Relative time
            const parts = scheduledTime.slice(1).split(' ');
            const amount = parseInt(parts[0]);
            const unit = parts[1]?.toLowerCase();
            
            scheduledDate = new Date();
            if (unit?.includes('min')) {
                scheduledDate.setMinutes(scheduledDate.getMinutes() + amount);
            } else if (unit?.includes('hour') || unit?.includes('hr')) {
                scheduledDate.setHours(scheduledDate.getHours() + amount);
            }
        } else {
            // Absolute time (HH:MM)
            const [hour, minute] = scheduledTime.split(':').map(Number);
            scheduledDate = new Date();
            scheduledDate.setHours(hour, minute, 0, 0);
            if (scheduledDate < new Date()) {
                scheduledDate.setDate(scheduledDate.getDate() + 1);
            }
        }

        const schedule = {
            id: scheduleId,
            user_id: userId,
            channel_id: channelId,
            scheduled_time: scheduledDate.toISOString(),
            recurring,
            active: true,
            created_at: new Date().toISOString()
        };

        this.schedules.push(schedule);
        this.saveSchedules();
        return scheduleId;
    }

    getUserSchedules(userId) {
        return this.schedules.filter(s => s.user_id === userId && s.active);
    }

    deleteSchedule(scheduleId, userId) {
        const schedule = this.schedules.find(s => s.id === scheduleId && s.user_id === userId);
        if (schedule) {
            schedule.active = false;
            this.saveSchedules();
            return true;
        }
        return false;
    }

    getActiveSchedules() {
        return this.schedules.filter(s => s.active);
    }
}

export const scheduler = new Scheduler();

File: managers/scheduler.py
"""
Scheduler untuk Scheduled Stay system
"""
import json
import os
from typing import Dict, Optional, List
from datetime import datetime, timedelta
import asyncio

SCHEDULES_DB_FILE = "schedules.json"


class Scheduler:
    """Manager untuk handle scheduled stays"""
    
    def __init__(self):
        self.schedules: List[Dict] = []
        self.running_tasks: Dict[int, asyncio.Task] = {}  # {schedule_id: task}
        self.load_schedules()
    
    def load_schedules(self):
        """Load schedules dari file"""
        if os.path.exists(SCHEDULES_DB_FILE):
            try:
                with open(SCHEDULES_DB_FILE, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.schedules = data.get("schedules", [])
            except Exception as e:
                print(f"⚠️  Error loading schedules: {e}")
                self.schedules = []
        else:
            self.schedules = []
    
    def save_schedules(self):
        """Save schedules ke file"""
        try:
            with open(SCHEDULES_DB_FILE, 'w', encoding='utf-8') as f:
                json.dump({"schedules": self.schedules}, f, indent=2, ensure_ascii=False, default=str)
        except Exception as e:
            print(f"⚠️  Error saving schedules: {e}")
    
    def create_schedule(self, user_id: int, channel_id: int, scheduled_time: str, 
                       recurring: Optional[str] = None) -> int:
        """Create new schedule"""
        schedule_id = len(self.schedules) + 1
        
        # Parse scheduled_time
        if scheduled_time.startswith("+"):
            # Relative time
            parts = scheduled_time[1:].split()
            if len(parts) == 2:
                amount = int(parts[0])
                unit = parts[1].lower()
                
                if unit in ["minute", "minutes", "min", "mins"]:
                    scheduled_datetime = datetime.now() + timedelta(minutes=amount)
                elif unit in ["hour", "hours", "hr", "hrs"]:
                    scheduled_datetime = datetime.now() + timedelta(hours=amount)
                else:
                    raise ValueError(f"Invalid time unit: {unit}")
            else:
                raise ValueError("Invalid relative time format")
        else:
            # Absolute time (HH:MM)
            try:
                hour, minute = map(int, scheduled_time.split(":"))
                scheduled_datetime = datetime.now().replace(hour=hour, minute=minute, second=0, microsecond=0)
                # If time already passed today, schedule for tomorrow
                if scheduled_datetime < datetime.now():
                    scheduled_datetime += timedelta(days=1)
            except:
                raise ValueError("Invalid time format. Use HH:MM or +X minutes/hours")
        
        schedule = {
            "id": schedule_id,
            "user_id": user_id,
            "channel_id": channel_id,
            "scheduled_time": scheduled_datetime.isoformat(),
            "recurring": recurring,  # "daily", "weekly", or None
            "active": True,
            "created_at": datetime.now().isoformat()
        }
        
        self.schedules.append(schedule)
        self.save_schedules()
        return schedule_id
    
    def get_user_schedules(self, user_id: int) -> List[Dict]:
        """Get all schedules for user"""
        return [s for s in self.schedules if s["user_id"] == user_id and s["active"]]
    
    def get_schedule(self, schedule_id: int) -> Optional[Dict]:
        """Get schedule by ID"""
        for schedule in self.schedules:
            if schedule["id"] == schedule_id:
                return schedule
        return None
    
    def delete_schedule(self, schedule_id: int, user_id: int) -> bool:
        """Delete schedule"""
        for schedule in self.schedules:
            if schedule["id"] == schedule_id and schedule["user_id"] == user_id:
                schedule["active"] = False
                self.save_schedules()
                
                # Cancel running task
                if schedule_id in self.running_tasks:
                    self.running_tasks[schedule_id].cancel()
                    del self.running_tasks[schedule_id]
                
                return True
        return False
    
    def edit_schedule(self, schedule_id: int, user_id: int, scheduled_time: str) -> bool:
        """Edit schedule time"""
        schedule = self.get_schedule(schedule_id)
        if not schedule or schedule["user_id"] != user_id:
            return False
        
        # Parse new time
        if scheduled_time.startswith("+"):
            parts = scheduled_time[1:].split()
            if len(parts) == 2:
                amount = int(parts[0])
                unit = parts[1].lower()
                
                if unit in ["minute", "minutes", "min", "mins"]:
                    scheduled_datetime = datetime.now() + timedelta(minutes=amount)
                elif unit in ["hour", "hours", "hr", "hrs"]:
                    scheduled_datetime = datetime.now() + timedelta(hours=amount)
                else:
                    return False
            else:
                return False
        else:
            try:
                hour, minute = map(int, scheduled_time.split(":"))
                scheduled_datetime = datetime.now().replace(hour=hour, minute=minute, second=0, microsecond=0)
                if scheduled_datetime < datetime.now():
                    scheduled_datetime += timedelta(days=1)
            except:
                return False
        
        schedule["scheduled_time"] = scheduled_datetime.isoformat()
        self.save_schedules()
        return True


# Global instance
scheduler = Scheduler()

